### PR TITLE
Add `DelayStrategy` to coroutine module

### PR DESCRIPTION
**Pull request by sdonn3**
In the process of fixing the websocket reconnection, I realized it would be nice to add the logic for ReconnectStrategy to our Tuulbox for reuse. It has been renamed DelayStrategy as it is open-ended and does not necessarily have to be reconnection-focused.

### CODE REVIEW

**Review by twyatt — COMMENTED** [review]
I like the idea of moving these to a public library, but we should add unit tests for it.

File: coroutines/src/commonMain/kotlin/DelayStrategy.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import kotlinx.coroutines.delay
```
Comment (twyatt) [inline coroutines/src/commonMain/kotlin/DelayStrategy.kt:1]: Missing package.

File: coroutines/src/commonMain/kotlin/DelayStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * Interface, using the strategy pattern, for an implementer to delay for periods of time.
+ */
+public interface DelayStrategy {
+    public suspend fun await(iteration: Int, elapsedMillis: Long)
+}
+
+/**
+ * DelayStrategy that, when await() is called, will delay for a fixed amount of time, to be determined by [delayMillis].
```
Comment (twyatt) [inline coroutines/src/commonMain/kotlin/DelayStrategy.kt:20]:
```suggestion
 * [DelayStrategy] that, when [await] is called, will delay for a fixed amount of time, specified by [delayMillis].
```

File: coroutines/src/commonMain/kotlin/DelayStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ */
+public class FixedDelay(private val delayMillis: Long) : DelayStrategy {
+    override suspend fun await(iteration: Int, elapsedMillis: Long) {
+        delay(delayMillis - elapsedMillis)
+    }
+}
+
+/**
+ * DelayStrategy with an exponentially increasing delay, calculated with [calculateExponentialBackoffDelay].
```
Comment (twyatt) [inline coroutines/src/commonMain/kotlin/DelayStrategy.kt:29]:
```suggestion
 * [DelayStrategy] with an exponentially increasing delay, calculated with [calculateExponentialBackoffDelay].
```

File: coroutines/src/commonMain/kotlin/DelayStrategy.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        delay(calculateExponentialBackoffDelay(iteration, elapsedMillis, minimumMillis, maximumMillis))
+    }
+}
+
+internal fun calculateExponentialBackoffDelay(
+    iteration: Int,
+    elapsedMillis: Long,
+    minimumMillis: Long,
+    maximumMillis: Long
+) = (E.pow(iteration / E).toLong() * 1000L).coerceIn(minimumMillis..maximumMillis) - elapsedMillis
```
Comment (QuantumRand) [inline coroutines/src/commonMain/kotlin/DelayStrategy.kt:47]: This exponential function is pretty specific to our needs. As a publicly consumed library, this should be more configurable. Perhaps `EponentialBackoff` can take lambdas for the base and the exponent?

Comment (sdonn3) [reply coroutines/src/commonMain/kotlin/DelayStrategy.kt:47]: While I agree that adding customizability probably makes sense in a public library, it feels awkward to allow for customization on that portion as the delay calculation is the vast majority of the utility of this function.

Maybe the `ExponentialBackoff` class should only exist on the app layer?

Comment (sdonn3) [reply coroutines/src/commonMain/kotlin/DelayStrategy.kt:47]: Alternatively, maybe we could find a more specific name to this particular algorithm?

Comment (twyatt) [reply coroutines/src/commonMain/kotlin/DelayStrategy.kt:47]: Maybe something along the lines of?:

```kotlin
private const val DEFAULT_MULTIPLIER = 2f
private const val DEFAULT_MAXIMUM_DELAY_MILLIS = Long.MAX_VALUE

/**
 * Exponential backoff implementation using the following formula:
 *
 * ```
 * delay = base * multiplier ^ retry
 * ```
 *
 * For example (using `base = 100` and `multiplier = 2`):
 *
 * | retry | delay |
 * |-------|-------|
 * |   1   |   100 |
 * |   2   |   200 |
 * |   3   |   400 |
 * |   4   |   800 |
 * |   5   |  1600 |
 * |  ...  |   ... |
 *
 * Inspired by:
 * [Exponential Backoff And Jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)
 */
class ExponentialBackoff(
    private val baseTimeMillis: Long,
    private val multiplier: Float = DEFAULT_MULTIPLIER,
    maximumDelayMillis: Long = DEFAULT_MAXIMUM_DELAY_MILLIS
) : BackoffStrategy {

    private val maximumMillis = maximumDelayMillis.toFloat()

    override fun getDelayMillis(retry: Int): Long =
        (baseTimeMillis * multiplier.pow(retry - 1))
            .coerceAtMost(maximumMillis)
            .toLong()
}
```

_Pulled from old app code._

Even have some tests to go along with it if you want to pull them in here:

<details>
<summary>Tests<summary>

```kotlin
class ExponentialBackoffTest {

    @Test
    fun defaultsWithMaximumDelayMillis_largeRetryNumber_adheresToMaximum() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100, maximumDelayMillis = 1_000L)
        assertEquals(1_000L, backoff.getDelayMillis(retry = 20))
    }

    @Test
    fun calculatedDelayResultsInInfinity_returnsMaximumDelayMillis() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100_000L, maximumDelayMillis = 1_000L)
        assertEquals(1_000L, backoff.getDelayMillis(retry = 100_000))
    }

    @Test
    fun firstRetry_hasDelayEqualToBaseTimeMillis() {
        val backoff = ExponentialBackoff(baseTimeMillis = 100)
        assertEquals(100, backoff.getDelayMillis(retry = 1))
    }

    @Test
    fun multipleRetries_increaseExponentially() {
        val baseTimeMillis = 100L
        val backoff = ExponentialBackoff(baseTimeMillis)

        var delay = baseTimeMillis
        val retries = 1..10
        retries.forEach { retry ->
            assertEquals(delay, backoff.getDelayMillis(retry), "Backoff retry #$retry")
            delay *= 2
        }
    }
}
```
</details>

Comment (QuantumRand) [reply coroutines/src/commonMain/kotlin/DelayStrategy.kt:47]: This is what I was thinking
```kotlin
public class ExponentialBackoff(
    private val minimumMillis: Long,
    private val maximumMillis: Long,
    private val base: (iteration: Int) -> Float,
    private val exponent: (iteration: Int) -> Float,
) : DelayStrategy {
    override suspend fun await(iteration: Int, elapsedMillis: Long) {
        val delay = base.invoke(iteration).pow(exponent.invoke(iteration).coerceIn(minimumMillis..maximumMillis) - elapsedMillis
        delay(delay)
    }
}
```

Comment (sdonn3) [reply coroutines/src/commonMain/kotlin/DelayStrategy.kt]: I ended up going with something similar to Travis' proposal. While offering more customization, I found the lambda parameters to largely increase the complexity and required understanding of the public API.

I think this should cover our bases for our Delay implementation

File: coroutines/src/commonTest/kotlin/DelayStrategyTest.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        withTimeout(10_000) {
+            delay(100L)
+            assertEquals(Triple('A', 0, 1_000L), outputFlow.value)
+            trigger.value = false
+            delay(100L)
+            assertEquals(Triple('B', 0, 2_000L), outputFlow.value)
+            delayJob.cancelAndJoin()
+        }
+    }
```
Comment (twyatt) [inline coroutines/src/commonTest/kotlin/DelayStrategyTest.kt:92]: Let me know what you think; as slightly more controllable test (without the need for `delay`s in order to wait for conditions to be met, which are notoriously flaky for testes):

```suggestion
    @Test
    @OptIn(ExperimentalTime::class)
    fun dynamicDelayStrategy_switchesBetweenStrategies_usingTrigger() = runTest {
        val trigger = MutableStateFlow(true)
        val timeMachine = TimeMachine()
        val delayA = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 1_000L)
        val delayB = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 2_000L)
        val dynamic = Dynamic(trigger, timeMachine) { if (it) delayA else delayB }

        val job = launch {
            dynamic.await(iteration = 0, elapsedMillis = 0L)
        }

        assertEquals(
            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 0L),
            actual = delayA.invocation.receive(),
        )
        assertEquals(expected = false, actual = job.isCompleted)

        timeMachine.advanceBy(milliseconds = 500L)
        trigger.value = false
        assertEquals(
            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 500L),
            actual = delayB.invocation.receive(),
        )
        assertEquals(expected = false, actual = job.isCompleted)

        timeMachine.advanceBy(3_000L)
    }
```

...expects the change in the subsequent comment to also be made.

File: coroutines/src/commonTest/kotlin/DelayStrategyTest.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    private class TestDelay(
+        private val onAwait: (Pair<Int, Long>) -> Unit
+    ) : DelayStrategy {
+        override suspend fun await(iteration: Int, elapsedMillis: Long) {
+            onAwait(iteration to elapsedMillis)
+            awaitCancellation()
+        }
+    }
+}
```
Comment (twyatt) [inline coroutines/src/commonTest/kotlin/DelayStrategyTest.kt:102]:
```suggestion
@OptIn(ExperimentalTime::class)
private class TimeMachine : TimeSource {

    private val elapsedMillis = MutableStateFlow(0L)

    fun advanceBy(milliseconds: Long) { elapsedMillis.update { value -> value + milliseconds } }
    suspend fun delayUntil(milliseconds: Long) { elapsedMillis.first { it >= milliseconds } }

    private class TimeMachineMark(private val parent: TimeMachine) : TimeMark() {
        override fun elapsedNow(): Duration = parent.elapsedMillis.value.toDuration(MILLISECONDS)
    }

    override fun markNow(): TimeMark = TimeMachineMark(this)
}

private class Invocation {
    data class Await(val iteration: Int, val elapsedMilliseconds: Long)
}

private class TimeMachineDelayStrategy(
    private val timeMachine: TimeMachine,
    private val delayUntilMilliseconds: Long,
) : DelayStrategy {

    val invocation = Channel<Invocation.Await>()

    override suspend fun await(iteration: Int, elapsedMillis: Long) {
        invocation.send(Invocation.Await(iteration, elapsedMillis))
        timeMachine.delayUntil(milliseconds = delayUntilMilliseconds)
    }
}
```

Comment (sdonn3) [reply coroutines/src/commonTest/kotlin/DelayStrategyTest.kt]: Yeah I was playing around with it quite a bit yesterday, and a `Channel` totally makes sense here. I like it!

File: coroutines/src/commonTest/kotlin/DelayStrategyTest.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        assertFalse(job.isCompleted)
+        timeMachine.advanceBy(milliseconds = 500L)
+        trigger.value = false
+        assertEquals(
+            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 500L),
+            actual = delayB.invocation.receive(),
+        )
+        assertFalse(job.isCompleted)
+        timeMachine.advanceBy(milliseconds = 3_000L)
+    }
```
Comment (twyatt) [inline coroutines/src/commonTest/kotlin/DelayStrategyTest.kt:75]: nit: whitespace would help with readability a bit.

```suggestion
    @Test
    @OptIn(ExperimentalTime::class)
    fun dynamicDelayStrategy_switchesBetweenStrategies_usingTrigger() = runTest {
        val trigger = MutableStateFlow(true)
        val timeMachine = TimeMachine()
        val delayA = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 1_000L)
        val delayB = TimeMachineDelayStrategy(timeMachine, delayUntilMilliseconds = 2_000L)
        val dynamic = Dynamic(trigger, timeMachine) { if (it) delayA else delayB }

        val job = launch {
            dynamic.await(iteration = 0, elapsedMillis = 0L)
        }
        assertEquals(
            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 0L),
            actual = delayA.invocation.receive(),
        )
        assertFalse(job.isCompleted)

        timeMachine.advanceBy(milliseconds = 500L)
        trigger.value = false
        assertEquals(
            expected = Invocation.Await(iteration = 0, elapsedMilliseconds = 500L),
            actual = delayB.invocation.receive(),
        )
        assertFalse(job.isCompleted)

        timeMachine.advanceBy(milliseconds = 3_000L)
    }
```

File: coroutines/src/commonTest/kotlin/DelayStrategyTest.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    private class TimeMachineDelayStrategy(
+        private val timeMachine: TimeMachine,
+        private val delayUntilMilliseconds: Long,
+    ) : DelayStrategy {
+        val invocation = Channel<Invocation>()
+        override suspend fun await(iteration: Int, elapsedMillis: Long) {
+            invocation.send(Invocation.Await(iteration, elapsedMillis))
+            timeMachine.delayUntil(milliseconds = delayUntilMilliseconds)
+        }
+    }
```
Comment (twyatt) [inline coroutines/src/commonTest/kotlin/DelayStrategyTest.kt:99]: There should be moved outside the test class, IMHO.
i.e. to the end of the file.

**Review by twyatt — APPROVED** [review]

**Review by cedrickcooke — APPROVED** [review]
Love the casual `TimeMachine`

**Pull request merged**